src/filters/EnergyFilter.C

```diff
@@ -0,0 +1,877 @@
+/********************************************************************/
+/*                  SOFTWARE COPYRIGHT NOTIFICATION                 */
+/*                             Cardinal                             */
+/*                                                                  */
+/*                  (c) 2021 UChicago Argonne, LLC                  */
+/*                        ALL RIGHTS RESERVED                       */
+/*                                                                  */
+/*                 Prepared by UChicago Argonne, LLC                */
+/*               Under Contract No. DE-AC02-06CH11357               */
+/*                With the U. S. Department of Energy               */
+/*                                                                  */
+/*             Prepared by Battelle Energy Alliance, LLC            */
+/*               Under Contract No. DE-AC07-05ID14517               */
+/*                With the U. S. Department of Energy               */
+/*                                                                  */
+/*                 See LICENSE for full restrictions                */
+/********************************************************************/
+
+#ifdef ENABLE_OPENMC_COUPLING
+#include "EnergyFilter.h"
+
+#include "openmc/tallies/filter_energy.h"
+
+registerMooseObject("CardinalApp", EnergyFilter);
+
+InputParameters
+EnergyFilter::validParams()
+{
+  auto params = FilterBase::validParams();
+  params.addClassDescription(
+      "A class which provides a thin wrapper around an OpenMC EnergyFilter. Energy bins "
+      "can either be manually specified in 'energy_boundaries' or picked from a list "
+      "provided in 'group_structure'.");
+  params.addParam<std::vector<Real>>(
+      "energy_boundaries",
+      "The energy boundaries to use to form energy bins. The boundaries must be provided "
+      "in ascending order in terms of energy.");
+  params.addParam<MooseEnum>(
+      "group_structure",
+      getEnergyGroupsEnum(),
+      "The energy group structure to use from a list of popular group structures.");
+  params.addParam<bool>(
+      "reverse_bins",
+      false,
+      "Whether the bins should be output in reverse order. This is useful for comparing "
+      "deterministic transport codes with Cardinal, where the convention is that energy "
+      "groups are arranged in descending order (i.e. group 1 is fast, group 2 is thermal).");
+
+  return params;
+}
+
+EnergyFilter::EnergyFilter(const InputParameters & parameters)
+  : FilterBase(parameters), _reverse_bins(getParam<bool>("reverse_bins"))
+{
+  if (isParamValid("energy_boundaries") && isParamValid("group_structure"))
+    mooseError("You have set both 'energy_boundaries' and 'group_structure'! Only one bin "
```

Comment:
same comment as above regarding switching `&&` to `==` to hit two birds with one stone regarding error checking.